core/src/main/java/jenkins/cli/ListNodesCommand.java

```diff
@@ -0,0 +1,578 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.cli;
+
+import hudson.Extension;
+import hudson.cli.CLICommand;
+import hudson.model.Computer;
+import hudson.model.Node;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.args4j.Option;
+
+/**
+ * CLI command to list Jenkins nodes with various filtering options such as status, labels, and executors.
+ * Supports output in plain text or JSON format, with an optional verbose mode for detailed information.
+ */
+@Restricted(NoExternalUse.class)
+@Extension
+public class ListNodesCommand extends CLICommand {
+    private static final Logger LOGGER = Logger.getLogger(ListNodesCommand.class.getName());
+
+    private static final String NL = System.lineSeparator(); // Newline character for consistent output formatting
+
+    /** Maximum length for labels input to prevent excessive input size. */
+    private static final int MAX_LABELS_LENGTH = 1000;
+
+    private record NodeInfo(
+        String displayName,
+        boolean isOnline,
+        boolean isTemporarilyOffline,
+        String offlineCauseReason,
+        Set<String> labels,
+        int numExecutors,
+        String osDescription,
+        String architecture,
+        String javaVersion,
+        String connectedSince,
+        String clockDifference
+    ) {}
+
+    /**
+     * Node status filter options.
+     */
+    enum Status {
+        ONLINE, OFFLINE, TEMP_OFFLINE, ALL
+    }
+
+    /**
+     * Output format options for node listing.
+     */
+    enum OutputFormat {
+        JSON, PLAIN
+    }
+
+    /**
+     * Label match mode: ANY label must match, or ALL labels must match.
+     */
+    enum LabelMatchMode {
+        ANY, ALL
+    }
+
+    // Command-line options
+
+    /**
+     * Filter nodes by status: ONLINE, OFFLINE (all offline nodes, including temporarily offline), TEMP_OFFLINE (temporarily offline nodes only), or ALL (default).
+     */
+    @Option(name = "-status", usage = "Filter nodes by status: ONLINE, OFFLINE (all offline nodes, including temporarily offline), TEMP_OFFLINE (temporarily offline nodes only), or ALL")
+    private Status status = Status.ALL;
+
+    /**
+     * Filter nodes by labels, using a comma-separated list (no spaces) (e.g., label1,label2,label3).
+     */
+    @Option(name = "-labels", usage = "Filter nodes by labels (comma-separated, no spaces) (e.g., label1,label2,label3)", metaVar = "label1,label2,...")
```

Comment:
I think it would be better if we can use label expressions. That is what most people know as you already use it when you assign a job or a pipeline to an agent. This us much more powerful and makes the `-labels-mode` option obsolete